### PR TITLE
Tk refactor+Remote tool use

### PR DESCRIPTION
**Pull request by Crazylemon64**
TK can now use tools on tiles adjacent to the object, and also is more able to use RCDs, manipulate stacks, and now works more consistently (less cloned/duped references) when interacting with certain objects.

With stronger atmos now a factor, TK has greater applications in construction - in that when you're not adjacent to a vacuum gradient, you can more easily build without being flung around.

Also adds an `/obj/item/proc/override_throw(mob/user)` proc that lets you override the throwing behavior on an individual object - return `TRUE` from that proc if you want to do so.

There is also now `/obj/item/proc/is_equivalent(obj/item/I)` - this is used to determine whether two items can be considered "the same" - important for wrapper items like the robot grasper, or TK as in this PR.

:cl:Crazylemon
rscadd: TK can now manipulate adjacent tiles. Use throw mode to override this if you want to move an object a single tile.
rscadd: TK can now remotely manipulate item stacks.
rscadd: TK can now remotely operate an RCD
fix: TK now behaves more consistently regarding interacting with various objects remotely
/:cl:

### CODE REVIEW

**Comment by SamHPurp** [comment]
As we were talking about on Discord, I think this is really cool. 👍

**Comment by Anasari** [comment]
Stealth thumb up

**Comment by fludd12** [comment]
Does this mean TK can work on computers now? I'd be so excited if so.

**Comment by Anticept** [comment]
That could mean TK is both the most awesome, and at the same time the worst, thing to happen, if you can control things like computers remotely. There's a lot of damage you could potentially do on terminals that don't use access requirements like the R&D terminal.

Make sure it leaves fingerprints for admin tracing!

**Comment by taukausanake** [comment]
Is this going to inherent any problem similar to using or doing things when you physically shouldn't be able to? Like taking things that are on the other side of glass/in areas physically closed off from another and such?

**Comment by Crazylemon64** [comment]
I'm pretty sure you can already use computers from a distance with TK. This does not change that.

And no, this does not allow bending reality - it checks for the distance of the object from the target, not from you.

**Comment by fludd12** [comment]
Using computers bring s up the screen (sometimes?) but acts like you walked away, greying out buttons, at least last I checked.

**Comment by SamHPurp** [comment]
I assume it check adjacent and closes you

**Comment by Crazylemon64** [comment]
It allows you to interact with machines in situations where the eye would be orange (updates only, no interaction) - I made no change made there.

**Comment by Anticept** [comment]
Just checked.

All computers have a 2 tile interaction range with TK. Just enough to be Trouble, even through glass.

File: code/_onclick/telekinesis.dm [hunk 1]
```diff
@@ -75,104 +69,132 @@ var/const/tk_maxrange = 15
...
+		release_object()
+	// Focus is null now
+	host = null
+	return ..()
 
-	dropped(mob/user as mob)
-		if(focus && user && loc != user && loc != user.loc) // drop_item() gets called when you tk-attack a table/closet with an item
-			if(focus.Adjacent(loc))
-				focus.loc = loc
+/obj/item/tk_grab/dropped(mob/user as mob)
```
Comment (Fox-McCloud) [inline code/_onclick/telekinesis.dm]: please kill off all `as mob` in the files you edit.

File: code/_onclick/telekinesis.dm [hunk 2]
```diff
@@ -75,104 +69,132 @@ var/const/tk_maxrange = 15
...
 
-		qdel(src)
-		return
+	qdel(src)
 
 
 	//stops TK grabs being equipped anywhere but into hands
-	equipped(var/mob/user, var/slot)
-		if( (slot == slot_l_hand) || (slot== slot_r_hand) )	return
+/obj/item/tk_grab/equipped(var/mob/user, var/slot)
```
Comment (Fox-McCloud) [inline code/_onclick/telekinesis.dm]: and implied `var`: `/obj/item/tk_grab/equipped(mob/user, slot)`

File: code/_onclick/telekinesis.dm [hunk 3]
```diff
@@ -75,104 +69,132 @@ var/const/tk_maxrange = 15
...
+	if( (slot == slot_l_hand) || (slot== slot_r_hand) )	return
```
Comment (Fox-McCloud) [inline code/_onclick/telekinesis.dm]: new line.

File: code/_onclick/telekinesis.dm [hunk 4]
```diff
@@ -75,104 +69,132 @@ var/const/tk_maxrange = 15
...
+/obj/item/tk_grab/attack_self(mob/user as mob)
+	if(focus)
+		focus.attack_self_tk(user)
+
+/obj/item/tk_grab/override_throw(mob/user, atom/target)
+	afterattack(target, user)
+	return TRUE
+
+/obj/item/tk_grab/afterattack(atom/target as mob|obj|turf|area, mob/living/user as mob|obj, proximity, params)//TODO: go over this
+	if(!target || !user)	return
```
Comment (Fox-McCloud) [inline code/_onclick/telekinesis.dm]: v

File: code/_onclick/telekinesis.dm [hunk 5]
```diff
@@ -75,104 +69,132 @@ var/const/tk_maxrange = 15
...
+	if(last_throw+3 > world.time)	return
```
Comment (Fox-McCloud) [inline code/_onclick/telekinesis.dm]: new line.

File: code/_onclick/telekinesis.dm [hunk 6]
```diff
@@ -75,104 +69,132 @@ var/const/tk_maxrange = 15
...
+/obj/item/tk_grab/attack(mob/living/M as mob, mob/living/user as mob, def_zone)
+	return
+
+/obj/item/tk_grab/is_equivalent(obj/item/I)
+	. = ..()
+	if(!.)
+		return I == focus
 
+/obj/item/tk_grab/proc/focus_object(var/obj/target, var/mob/user)
+	if(!istype(target,/obj))	return//Cant throw non objects atm might let it do mobs later
```
Comment (Fox-McCloud) [inline code/_onclick/telekinesis.dm]: new line.

File: code/modules/mob/mob.dm [hunk 7]
```diff
@@ -13,6 +13,9 @@
 	for(var/mob/dead/observer/M in following_mobs)
 		M.following = null
 	following_mobs = null
+	for(var/I in tkgrabbed_objects)
+		qdel(tkgrabbed_objects[I])
```
Comment (Fox-McCloud) [inline code/modules/mob/mob.dm]: We have a define for this, I believe.

**Comment by Crazylemon64** [comment]
@Fox-McCloud did your stuff

**Pull request merged**